java-util/src/main/java/io/druid/java/util/common/concurrent/SameThreadExecutorService.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.java.util.common.concurrent;
+
+import java.util.List;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A simple class that implements the ExecutorService interface, but runs the code on a call to submit
+ */
+public class SameThreadExecutorService extends AbstractExecutorService
+{
+
+  @Override
+  public void shutdown()
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public List<Runnable> shutdownNow()
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean isShutdown()
+  {
+    return false;
+  }
+
+  @Override
+  public boolean isTerminated()
+  {
+    return false;
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException
+  {
+    final long nanos = TimeUnit.NANOSECONDS.convert(timeout, unit);
+    final long millis = TimeUnit.MILLISECONDS.convert(timeout, unit);
+    final int sleepNanos = (int) (nanos - millis * 1_000_000L);
+    Thread.sleep(millis, sleepNanos);
```

Comment:
I don't think sleeping makes much sense here, either return immediately or throw an UnsupportedOperationException.

Reply:
The contract says it will wait, so I wait

Reply:
`awaitTermination()` should be called only after `shutdown()` or `shutdownNow()`, which throw UnsupportedOperationException, so it's even more reasonable to throw it from `awaitTermination()`

Reply:
The interface is unambiguous : https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html#awaitTermination-long-java.util.concurrent.TimeUnit- 

```
Blocks until all tasks have completed execution after a shutdown request, or the timeout occurs, or the current thread is interrupted, whichever happens first.
```

Which is what this implementation does.

Reply:
Javadocs of `shutdown()` and `shutdownNow()` also don't give an option to throw an UnsupportedOperationException. Yet when `awaitTermination()` is called on `SameThreadExecutorService`, it should be a programming bug, so better to report it rather than silence.

Reply:
`UnsupportedOperationException` is an unchecked exception, it does not need to be declared

Reply:
Yes, and that is why there is no problem with throwing it from the method, although the interface doesn't specify. I mean, if we throw UOE from `shutdown()` and `shutdownNow()`, rather than make those methods no-ops, it makes just as much sense to throw UOE from `awaitTermination()` as well. And so to keep them consistent.

Reply:
@drcrallen the PR looks good to me, except this thing, but it's not blocking for merge. However I still think it should be changed.